### DISCUSSION TITLE
Example of insertion of many values, where the `Params` is a list/vector of non-tuple?

### DISCUSSION THREAD

**Discussion by magthe** [nikita-volkov/hasql#168] [Q&A] [answered]
The example [in the docs](https://hackage.haskell.org/package/hasql-1.8.1.1/docs/Hasql-Statement.html#g:2) is clear and understandable:

```
insertMultipleLocations :: Statement (Vector (UUID, Double, Double)) ()
insertMultipleLocations =
  Statement sql encoder decoder True
  where
    sql = "insert into location (id, x, y) select * from unnest ($1, $2, $3)"
    encoder =
      Data.Vector.unzip3 >$<
        Contravariant.Extras.contrazip3
          (Encoders.param $ Encoders.nonNullable $ Encoders.foldableArray $ Encoders.nonNullable Encoders.uuid)
          (Encoders.param $ Encoders.nonNullable $ Encoders.foldableArray $ Encoders.nonNullable Encoders.float8)
          (Encoders.param $ Encoders.nonNullable $ Encoders.foldableArray $ Encoders.nonNullable Encoders.float8)
    decoder = Decoders.noResult
```

However, I'm struggling with finding a way to adapt to cases where the params aren't a vector/list of tuples. For instance, something like this

```
data MyType = MyType { id_ :: UUID, x :: Double, y :: Double}

insertMultipleLocations :: Statement (Vector MyType) ()
insertMultipleLocations =
  Statement sql encoder decoder True
  where
    sql = "insert into location (id, x, y) select * from unnest ($1, $2, $3)"
    encoder = ???
    decoder = Decoders.noResult
```

Do you have any good patterns for writing the encoder?

I could write a function `myTypeToTuple :: MyType -> (UUID, Double, Double)` and compose it with `unzip3`, but in my case I have a type with 19 fields (and likely to grow) so it's a bit of a pain to do.

**Comment by nikita-volkov** [comment, marked as answer]
Now that we have the pipelining mode supported, you can simply traverse with a single-row insert statement from inside of the `Pipeline` context. It should be comparable in terms of performance.

**Comment by magthe** [comment]
Thanks @nikita-volkov . I still haven't found the time to try this out, but it looks nice. I'll be back with a followup ASAP, but for now I'm closing this as answered.

**Comment by magthe** [comment]
Yes, that worked beautifully. Thanks :pray: 

The only thing I noticed was that the [example](https://hackage.haskell.org/package/hasql-1.8.1.3/docs/Hasql-Pipeline.html) uses `forM`, but `Pipeline` isn't a `Monad`. I suspect it's a spelling mistake only, `for` works.

**Reply by nikita-volkov** [reply]
Thanks for the `forM` mention. I've updated the docs.
